aten/src/ATen/native/cuda/TriangularOps.cu

```diff
@@ -106,5 +106,149 @@ Tensor& triu_cuda_out(Tensor &result, const Tensor& self, int64_t k) {
   return triu_tril_cuda_template<true>(result, self, k, "triu");
 }
 
-}  // namespace native
-}  // namespace at
+// Copy the kth diagonal of a matrix B to a vector A.
+template <typename scalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(1024)
+#endif
+__global__ void copy_from_diagonal_kernel(
+    scalar_t* a,
+    scalar_t* b,
+    std::ptrdiff_t start,
+    std::ptrdiff_t size,
+    std::ptrdiff_t strideSum,
+    std::ptrdiff_t strideA) {
+  for (std::ptrdiff_t linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < size;
+       linearIndex += gridDim.x * blockDim.x) {
+    const std::ptrdiff_t bOffset = start + strideSum * linearIndex;
+    a[strideA * linearIndex] = b[bOffset];
+  }
+}
+
+// Copy vector B to the kth diagonal of a matrix A
+template <typename scalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(1024)
+#endif
+__global__ void copy_to_diagonal_kernel(
+    scalar_t* a,
+    scalar_t* b,
+    std::ptrdiff_t start,
+    std::ptrdiff_t size,
+    std::ptrdiff_t strideSum,
+    std::ptrdiff_t strideB) {
+  for (std::ptrdiff_t linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < size;
+       linearIndex += gridDim.x * blockDim.x) {
+    const std::ptrdiff_t aOffset = start + strideSum * linearIndex;
+    a[aOffset] = b[strideB * linearIndex];
+  }
+}
+
+template <typename scalar_t>
+Tensor& apply_diag(Tensor& result, const Tensor& self, int64_t dimension) {
+  TORCH_CHECK(
+      self.dim() == 1 || self.dim() == 2, "matrix or a vector expected");
+
+  TensorArg result_arg{result, "result", 1};
+  TensorArg self_arg{self, "self", 2};
+  checkAllSameGPU("diag", {result_arg, self_arg});
+  checkSameType("diag", result_arg, self_arg);
+
+  auto check_zero_strided = [](const IntArrayRef& strides) {
+    auto result = std::find(strides.begin(), strides.end(), 0);
+    if (result != strides.end()) {
+      return true;
+    }
+    return false;
+  };
+
+  int nDimension = self.dim();
+  if (nDimension == 2) {
+    auto self_stride_0 = self.stride(0);
+    auto self_stride_1 = self.stride(1);
+
+    int sz;
+    if (dimension > 0) {
+      sz = std::min(self.size(0), self.size(1) - dimension);
+    } else {
+      sz = std::min(self.size(0) + dimension, self.size(1));
+    }
+
+    result.resize_({sz});
+    if (sz > 0) {
+      if (check_zero_strided(result.strides())) {
+        result = result.contiguous();
```

Comment:
it is better to error out if `out` is self-overlapping (that would be consistent with behavior for pointwise operations)